src/main/java/slimeknights/tconstruct/smeltery/recipe/AbstractCastingRecipe.java

```diff
@@ -0,0 +1,173 @@
+  @Getter
+  private final Ingredient cast;
+  /** FluidStack for recipe (fluid and amount) */
+  @Getter
+  private final FluidStack fluid;
+  private final ItemStack result;
+  /** How long it takes to cool off */
+  @Getter
+  protected final int coolingTime;
+  /** Whether the cast is consumed when cast */
+  @Getter
+  private final boolean consumed;
+  private final boolean switchSlots;
+
+  public AbstractCastingRecipe(IRecipeType<?> typeIn, ResourceLocation idIn, String groupIn,
+                               @Nullable Ingredient ingredient, @Nonnull FluidStack fluidIn, ItemStack result,
+                               int coolingTime, boolean consumed, boolean switchSlots) {
+    this.type = typeIn;
+    this.id = idIn;
+    this.group = groupIn;
+    this.cast = ingredient;
+    this.fluid = fluidIn;
+    this.result = result;
+    this.coolingTime = coolingTime;
+    this.consumed = consumed;
+    this.switchSlots = switchSlots;
+  }
+
+  @Override
+  public boolean matches(ICastingInventory inv, World worldIn) {
+    return this.cast.test(inv.getStack()) && this.fluid.getFluid() == inv.getFluid();
+  }
+
+  @Override
+  public ItemStack getCraftingResult(ICastingInventory inv) {
+    return result;
+  }
+
+  @Override
+  public boolean canFit(int width, int height) {
+    return true;
+  }
+
+  @Override
+  public ItemStack getRecipeOutput() {
+    return this.result;
+  }
+
+  @Override
+  public ResourceLocation getId() {
+    return this.id;
+  }
+
+  @Override
+  public abstract IRecipeSerializer<?> getSerializer();
+
+  @Override
+  public IRecipeType<?> getType() {
+    return type;
+  }
+
+  /**
+   * Does this recipe's output get put into the input slot
+   * @return  Whether to switch to input after casting
+   */
+  public boolean switchSlots() {
+    return this.switchSlots;
+  }
+
+  /**
+   * Keeps casting recipes out of the recipe book.
+   */
+  @Override
+  public boolean isDynamic() {
+    return true;
+  }
+
+  public static class Serializer<T extends AbstractCastingRecipe> extends ForgeRegistryEntry<IRecipeSerializer<?>>
+    implements IRecipeSerializer<AbstractCastingRecipe> {
+    private final Serializer.IFactory<T> factory;
+
+    public Serializer(Serializer.IFactory<T> factoryIn) {
+      this.factory = factoryIn;
+    }
+
+    @Override
+    public T read(ResourceLocation recipeId, JsonObject json) {
+      Ingredient cast = Ingredient.EMPTY;
+      String s = JSONUtils.getString(json, "group", "");
+      boolean consumed = false;
+      boolean switchSlots = JSONUtils.getBoolean(json, "switch_slots", false);
+      if (json.has("cast")) {
+        JsonElement jsonelement = JSONUtils.getJsonObject(json, "cast");
+        cast = Ingredient.deserialize(jsonelement);
+        consumed = JSONUtils.getBoolean(json, "cast_consumed", false);
+      }
+
+      if (!json.has("fluidstack"))
+        throw new JsonSyntaxException("Missing fluid input definition!");
```

Comment:
This is redundant to the line below, `JSONUtils.getWhatever` methods all throw if the whatever is missing.